sys/mips/include/atomic.h

```diff
@@ -882,6 +992,229 @@ atomic_fetchadd_64(__volatile uint64_t *p, uint64_t v)
 }
 #endif
 
+
+#ifdef __CHERI_PURE_CAPABILITY__
+/*
+ * CHERI-only variants to perform atomic operations on pointers.
+ */
+
+static __inline void
+atomic_set_ptr(volatile uintptr_t *p, uintptr_t v)
+{
+	uintptr_t temp;
+
+	__asm __volatile (
+		"1:\n\t"
+		"cllc		%0, %1\n\t"	/* load old value */
+		"cgetoffset	$t0, %0\n\t"	/* calculate new value */
+		"cgetoffset	$t1, %2\n\t"
+		"or		$t0, $t1, $t0\n\t"
+		"csetoffset	%0, %0, $t0\n\t"
+		"cscc		$t0, %0, %1\n\t" /* attempt to store */
+		"beqz		$t0, 1b\n\t"	/* spin if failed */
+		"nop\n\t"			/* delay slot */
+		: "=&r" (temp), "+C" (p)
+		: "r" (v)
+		: "t0", "t1", "memory");
+}
+
+static __inline void
+atomic_clear_ptr(volatile uintptr_t *p, uintptr_t v)
+{
+	uintptr_t temp;
+	v = ~v;
+
+	__asm __volatile (
+		"1:\n\t"
+		"cllc		%0, %1\n\t"	/* load old value */
+		"cgetoffset	$t0, %0\n\t"	/* calculate new value */
+		"cgetoffset	$t1, %2\n\t"
+		"and		$t0, $t1, $t0\n\t"
+		"csetoffset	%0, %0, $t0\n\t"
+		"cscc		$t0, %0, %1\n\t" /* attempt to store */
+		"beqz		$t0, 1b\n\t"	/* spin if failed */
+		"nop\n\t"			/* delay slot */
+		: "=&r" (temp), "+C" (p)
+		: "r" (v)
+		: "t0", "t1", "memory");
+}
+
+static __inline void
+atomic_add_ptr(volatile uintptr_t *p, uintptr_t v)
+{
+	uintptr_t temp;
+
+	__asm __volatile (
+		"1:\n\t"
+		"cllc		%0, %1\n\t"	/* load old value */
+		"cgetoffset	$t0, %2\n\t"	/* calculate new value */
```

Comment:
I'd go for addr here even though it shouldn't make a difference (you'd hope v is NULL-derived).

Reply:
Technically this should be the case for `subtract_ptr` as well